lib/ansible/modules/cloud/smartos/vmadm.py

```diff
@@ -0,0 +1,600 @@
+
+    return changed, vm_uuid
+
+
+def vmadm_create_vm(module, payload_file):
+    """Create a new VM using the provided payload."""
+    cmd = 'vmadm create -f {0}'.format(payload_file)
+
+    return module.run_command(cmd)
+
+
+def set_vm_state(module, vm_uuid, vm_state):
+    p = module.params
+
+    # Check if the VM is already in the desired state.
+    state = get_vm_prop(module, vm_uuid, 'state')
+    if state and (state == vm_state):
+        return None
+
+    # Lookup table for the state to be in, and which command to use for that.
+    # vm_state: [vmadm commandm, forceable?]
+    cmds = {
+        'stopped': ['stop', True],
+        'running': ['start', False],
+        'deleted': ['delete', True],
+        'rebooted': ['reboot', False]
+    }
+
+    if p['force'] and cmds[vm_state][1]:
+        force = '-F'
+    else:
+        force = ''
+
+    cmd = 'vmadm {0} {1} {2}'.format(cmds[vm_state][0], force, vm_uuid)
+
+    (rc, stdout, stderr) = module.run_command(cmd)
+
+    match = re.match('^Successfully.*', stderr)
+    if match:
+        return True
+    else:
+        return False
+
+
+def create_payload(module, uuid):
+    """Create the JSON payload (vmdef) and return the filename."""
+
+    p = module.params
+
+    # Filter out the few options that are not valid VM properties.
+    module_options = ['debug', 'force', 'state']
+    vmattrs = filter(lambda prop: prop not in module_options, p)
+
+    vmdef = {}
+
+    for attr in vmattrs:
+        if p[attr]:
+            vmdef[attr] = p[attr]
+
+    try:
+        vmdef_json = json.dumps(vmdef)
+    except Exception as e:
+        module.fail_json(msg='Could not create valid JSON payload: {0}'.format(e))
+
+    # Create the temporary file that contains our payload, and set tight
+    # permissions for it may container sensitive information.
+    try:
+        fname = tempfile.mkstemp(uuid)[1]
+        fh = open(fname, 'w')
+        os.chmod(fname, 0o400)
+        fh.write(vmdef_json)
+        fh.close()
+    except Exception as e:
+        module.fail_json(msg='Could not save JSON payload as {0}: {1}'.format(fname, e))
+
+    return fname
+
+
+def vm_state_transition(module, uuid, vm_state):
+    ret = set_vm_state(module, uuid, vm_state)
+
+    # Whether the VM changed state.
+    if ret is None:
+        return False
+    elif ret:
+        return True
+    else:
+        module.fail_json(msg='Failed to set VM {0} to state {1}'.format(uuid, vm_state))
+
+
+def is_valid_uuid(uuid):
+    if re.match('^[0-9a-f]{8}-([0-9a-f]{4}-){3}[0-9a-f]{12}$', uuid, re.IGNORECASE):
+        return True
+    else:
+        return False
+
+
+def validate_uuids(module):
+    # Perform basic UUID validation.
+    for u in [['uuid', module.params['uuid']],
+              ['image_uuid', module.params['image_uuid']]]:
+        if u[1] and u[1] != '*':
+            if not is_valid_uuid(u[1]):
+                module.fail_json(msg='{0} is not a valid UUID for {1}'.format(u[1], u[0]))
```

Comment:
not a requirement, but would be nice to validate all IDs before returning an error (in case more than one is wrong)